### ISSUE TITLE
Support for 1D LUTs

### ISSUE THREAD

**Issue by donmccurdy** [pmndrs/postprocessing#368]
For [three-filmic](https://github.com/donmccurdy/three-filmic), I needed a small 1D "shaper" LUT in addition to the larger 3D LUT. The 1D version is really simple, implemented here as [LUT1DEffect.ts](https://github.com/donmccurdy/three-filmic/blob/main/src/effects/LUT1DEffect.ts). Would you be interested in having support for 1D LUTs in this repository? If so, I'd be happy to make a PR. Should it be part of LUTEffect, or separate the two as (for example) LUT1DEffect and LUT3DEffect?

Related — I found that LUTCubeLoader does not currently support 1D `.cube` LUTs. That'd be pretty easy to add as well, if we're interested. I wasn't sure why LUTCubeLoader is forked from the three.js version though — is there an important difference between the two? For bandwidth reasons, I'm planning to eventually switch three-filmic from plaintex `.cube` LUTs to compressed `.ktx2` data textures.

### Additional context

- https://github.com/pmndrs/postprocessing/issues/124
- https://github.com/pmndrs/postprocessing/issues/231

**Comment by vanruesc** [comment]
> Would you be interested in having support for 1D LUTs in this repository?

I don't think I've used 1D LUTs yet.. what are their use cases? Could this be added to `LUTEffect` as a supporting setting for the 3D LUT or is a 1D LUT a standalone kind of thing?

If it's mostly unrelated to 3D LUTs, then it should probably be its own effect, especially since `LUTEffect` focuses on 3D LUTs. Renaming that to `LUT3DEffect` would be fine.

> I wasn't sure why LUTCubeLoader is forked from the three.js version though — is there an important difference between the two?

The loaders were added together with `LUTEffect` and `Lookup3DTexture` when @gkjohnson contributed the 
`LUTPass` to `three` via https://github.com/mrdoob/three.js/pull/20558. There is no difference in functionality when comparing the .cube and .3dl loaders with the ones in `three`, but the loaders from `postprocessing` are a little bit faster: https://github.com/pmndrs/postprocessing/issues/231#issuecomment-753608834.

**Comment by vanruesc** [comment]
Btw: it's not required to call `setChanged` [here](https://github.com/donmccurdy/three-filmic/blob/27d1b7a7dbc8d06ac53acdf55616ea43dfb9d5fa/src/effects/LUT1DEffect.ts#L43) - that's only necessary when you add/remove uniforms or change preprocessor macros and WebGL extensions.

**Comment by donmccurdy** [comment]
> I don't think I've used 1D LUTs yet.. what are their use cases?

I'm porting an [OpenColorIO configuration](https://opencolorio.org/) and only partially understand the details, but I think the main advantage is that if you don't need one channel to affect another channel, a 1D LUT is more compact, precise, and efficient. Affinity offers a helpful guide, [1D vs 3D LUTs](https://affinityspotlight.com/article/1d-vs-3d-luts/), which says:

_Camera colour space conversions, especially through [OpenColorIO](http://opencolorio.org/) configurations, will typically use a 1D LUT as a ‘shaper’—usually to convert to linear. This then reduces the precision requirement of the 3D LUT, meaning a smaller cube size such as 17x17x17 can be used._

In this case I'm using a 1D LUT to adjust contrast, and a 3D LUT for filmic desaturation.

> Could this be added to LUTEffect as a supporting setting for the 3D LUT or is a 1D LUT a standalone kind of thing?

Yes — the 1D implementation is much simpler than the 3D implementation. I lean toward separating LUT1DEffect from LUT3DEffect so we don't have to deal with properties that are only valid for 3D (tetrahedralInterpolation) or have different types for 1D vs 3D  (scale, offset).

> There is no difference in functionality when comparing the .cube and .3dl loaders with the ones in three...

It looks like LUTEffect requires a LookupTexture, not a Data3DTexture ... Is there value in keeping customized LUTCubeLoader and LookupTexture types in this repository?  Or would you just as soon see those changes land back in three.js? Since `.cube` is not a particularly compact format I'd been thinking that storing LUTs as `.exr` or `.ktx2` textures would be more efficient (at least for my project), but I don't necessarily want to fork EXRLoader and KTX2Loader to output postprocessing-specific LookupTexture types, either. 🤔

> Btw: it's not required to call setChanged [here](https://github.com/donmccurdy/three-filmic/blob/27d1b7a7dbc8d06ac53acdf55616ea43dfb9d5fa/src/effects/LUT1DEffect.ts#L43)

Will make that change - thanks!

**Comment by donmccurdy** [comment]
tl;dr — I think my proposal would be:

1. Rename LUTEffect → LUT3DEffect
2. Add LUT1DEffect
3. Extend LUTCubeLoader to support 1D Adobe `.cube` LUTs
4. Contribute changes to LUTCubeLoader back to three.js¹

I'll dig into compressing LUTs with KTX2 later, I don't think that would affect anything in this repository directly.

¹ If this is OK with you? Code in three.js would be under MIT License. Not sure if you'd want to switch to the three.js version or keep a fork here.

**Comment by vanruesc** [comment]
Thanks for the additional information!

> In this case I'm using a 1D LUT to adjust contrast, and a 3D LUT for filmic desaturation.

I'm not sure what the advantage of a 1D LUT would be compared to using a `BrightnessContrastEffect` in this case, but maybe it's more about the workflow/portability aspects since it enables artists to tune a 1D LUT externally, then load it and apply it without having to tweak settings of foreign concepts like effects.

> It looks like LUTEffect requires a LookupTexture, not a Data3DTexture ... Is there value in keeping customized LUTCubeLoader and LookupTexture types in this repository?

The `LUTEffect` only expects a `Texture`. It can deal with 2D `Texture` strips, `Data3DTexture` and `LookupTexture`. I'm fine with deprecating the loaders, even if their performance tweaks don't make it into `three`. `LookupTexture` must stay though, either here or in `three` because it does the following:
 - stores domain information
 - implements LUT upscaling via web worker
 - supports conversion of JPG/PNG LUT strips into 3D textures via `from()`
 - supports conversion of `FloatType` data into `UnsignedByteType` data and vice versa
 - supports conversion of 2D data into 3D data and vice versa

The loaders in `postprocessing` return a `LookupTexture` for convenience, but they could also return a `Data3DTexture` instead. Since `LookupTexture` is a `Texture`, it can `copy()` data from a `Data3DTexture` instance.

> ¹ If this is OK with you? Code in three.js would be under MIT License. Not sure if you'd want to switch to the three.js version or keep a fork here.

Sure, I don't mind either way.

**Comment by donmccurdy** [comment]
> I'm not sure what the advantage of a 1D LUT would be compared to using a BrightnessContrastEffect...

There are other ways to change contrast, e.g. curves with a smoother toe and shoulder. For this tone mapping stack I'm working in a filmic log encoding color space — the 1D LUT does contrast adjustments, _and_ converts from the log encoding to sRGB in one step.

**Comment by donmccurdy** [comment]
Related PRs to three.js:

- https://github.com/mrdoob/three.js/pull/24102
- https://github.com/mrdoob/three.js/pull/24129

In tests far, a 130kb .ktx2 file can represent the same 3D LUT (at same precision) as a 880kb .cube file, and it's much faster to parse. That's without even using KTX2's support for Zstandard compression, which could reduce it further. So I may just skip on trying to do any big changes to the three.js or pmndrs LUTCubeLoader and LUT3dlLoader classes at this point.

**Comment by donmccurdy** [comment]
Just as FYI, more details on the KTX2 LUT workflow described above —

https://www.donmccurdy.com/2022/08/31/compressing-luts-with-ktx2/
